feat: add 0021, file delivery over mqtt (#70)
emqx/eip: active/0021-transfer-files-over-mqtt.md

# MQTT based file transfer

## Changelog

* 2022-09-20: @qzhuyan, @id Initial draft

## Abstract

TODO:

## Motivation

> EMQX customers are asking for file transfer functionality from IoT devices to their cloud (primary use case), and from cloud to IoT devices over MQTT. Right now they are uploading files from devices via FTP or HTTPS (e.g. to S3), but it's not working well. For example, both FTP and HTTP servers usually struggle to keep with large number of simultaneous bandwith-intensive connections. Another issue is that uploading files over weak/spotty/unreliable network is not a good experience due to frequent connection reset and transfer restarts.

Reviewer: More reasons (better to transform the reasons into bullets)
- Adding features to use cases where MQTT is already in place, if other protocols are to be used, it's usually a very different tech stack to deal with. 
  - Another set of programming SDK
  - Different types of security aspects including authentication and authorisation

Device-to-cloud file transfer use cases:

* [CAN bus](https://en.wikipedia.org/wiki/CAN_bus) data
* Image taken by industry camera for Quality Assurance
* Large data file collected from forklift
* Video and audio data from truck cars, and video data captured by inbound unloading cameras
* Vehicle real-time logging data and messaging
* ML logs

Cloud-to-device file transfer use cases:

* AI/ML models
* Firmware upgrades

> With EMQX, Customers can transfer real-time IoT data (e.g. structured, as well as unstructured IoT, sensor, and industrial data), and will also support offline types of bulk data transfer (video, audio, images,  compressed log files, etc.), which, combined with EMQX's rules engine, will make it extremely easy for users to connect any IoT data to the cloud.

Reviewer: ```suggestion
With EMQX, Customers can transfer real-time IoT data (e.g. structured, as well as unstructured IoT, sensor, and industrial data), and will also support offline types of bulk data transfer (video, audio, images,  compressed log files, etc.), which, combined with EMQX's Rule Engine and Data Integration, will make it extremely easy for users to connect any IoT data to the cloud.
```

## Requirements

* Simultaneous transfer of MQTT and various types of files over a single network connection
* Separate data transfer and file transfer channels to achieve non-blocking transmission
  * Channel priority setting, specify the priority of different channels according to customer requirements
  * Channels can cooperate with each other to achieve overall flow control
  * Multiple channels can be specified for file transfer at the same time to increase file transfer parallelism
* Efficient transmission support in weak network
  * File transfer in weak networks such as Telematics and mobile

>   * Automatic file chunking, block-based granularity breakpoint transfer (or file-level breakpoint transfer?)

Reviewer 2: Can the block size configured by user? For example, 1024 is the default block size, and user can set to 2048 etc.

Author: I don't see why not.

  * Switch to TCP if QUIC is not available

> 

Reviewer 3: Q: is this supposed to be for 5.0?  4.x too?

## AWS IoT MQTT-based file delivery (reference design)

As an example of existing implementation we can look at AWS IoT Core [which provides functionality](https://docs.aws.amazon.com/iot/latest/developerguide/mqtt-based-file-delivery.html) to [deliver files to IoT devices](https://docs.aws.amazon.com/iot/latest/developerguide/mqtt-based-file-delivery-in-devices.html):

![AWS IoT MQTT-based file delivery Device --> Cloud](0021-assets/aws-mqtt-file-delivery.png)

## Design

In the first phase we are going to focus on limited scope:

* device-to-cloud file transfer only
* reusing existing TCP connection
* push only, i.e. broker cannot request file transfer, but device can start sending files any time
* device may not know total file size when starting the transfer

> 

Reviewer 3: Q: if a device uploads a file and, before it publishes to the `/fin` topic to finalize the transfer, it crashes and "forgets" that it has already uploaded something.  Would we have a periodic sweeper process that clears those orphan files / segments?

Author: server will have some max retention and will be killing unfinished segments

TODO: add this to the spec

### Requirements

* Based on the MQTT protocol. Use MQTT messages to encapsulate file bodies
* Transport layer TCP, TCP/TLS and QUIC (TCP fallback) multiplexing connections with other service topics​
* High concurrency (multiple uploads in progress)
* MQTT Client single/bi-directional transfer (active file upload, passive file download)
* Fault tolerance: disconnected retransmission (sender side)
* Fragmentation: (multi parts, fragments)
* Multi-channel: multiple TCP connections, QUIC multi-streams
* Cancellation: Both parties can pause or cancel the transfer, and an event notification is required when the transfer is complete. (How to notify? What is the target of the notification?)
* Minimal code changes on Client side, SDK support available. (low code/no code)
* Minimize data retransmission, and reduce network pressure and bandwidth usage.
* RestAPI interface Query transfer status
* EMQX local incomplete file cache
* Client API should be easy to use: easy to migrate from existing applications (ftp, nfs, s3)
* File body Encryption and decryption (with external key management)?

### Out of scope

* Non-reliable transmission is not supported
* No support for one-to-many transfers

> * No distributed cache storage

Reviewer 3: This means that the device must send all the payloads to the same broker, right?  That would require correct load-balancer configuration for the session to be sticky, I guess?

Author: Brokers will be storing segments in local cache and gather segments when file transfer is finalized based on information in the session.
This is implementation specific, we probably don't want to add this into the spec.

* Broker does not aware of interrupted transfer, It will be client-side implementation

### Limitations

* File transfer is PUSH mechanism but not support POLL, the receiving side cannot request to initiate the transfer

> decisions of the size of a slice cannot be as natural as streaming.

Reviewer 3: Should this be a separate item?  Or is it indeed part of the previous phrase in the previous line?

* Due to network bandwidth limitations, it may happen that multiple files are transferred in parallel but all fail to complete the transfer in the desired time.

> * Do not get too hung up on the length of the topic and do not take advantage of "topic alias" (MQTT 5.0)

Reviewer: this protocol is **over** MQTT, so the use or no-use of "topic alias" should not be in this level's discussion.

* Performance may be limited by disk IO (network in speed > disk out speed)

> * Overwrite existing file SEG

Reviewer: ```suggestion
* At least once delivery, resent (existing) file segments will be overwritten.
* ```

* Linux filename length 255 bytes. path_max 4096

> * EMQX terminates file transfers, file message MQTT forwarding (as proxy) is not required.

Reviewer: ```suggestion
* EMQX terminates file transfers, file message MQTT forwarding (as proxy) is not supported.
```

> * QUIC in EMQX does not support HTTP/3 protocol, old client supports HTTP upload, how to smooth migration?

Reviewer: We want to have a clear boundary/scope for this project.
Move the uncertain parts to "Discussion"
```suggestion
* When using MQTT over QUIC as transport, how to take advantage of QUIC protocol (or HTTP3) is not out of the scope of this EIP.
```

### Protocol design

* Maximum bandwidth utilization, high signal-to-noise ratio (DATA/ MQTT HEADER)
* Asynchronous stateless non-continuous transfer, sliceable, messy (retransmission after disconnection) (parallel transfer of multiple files)
* Fault tolerance.
  * Incomplete packets Disconnected.
  * Transmission can be paused or resumed for a long time
  * Interactive completions required???
* Local caching of file contents

> * Utilizes only MQTT topic and payload (Payload Format Indicator?). No header changes.

Reviewer: ```suggestion
* Utilizes only MQTT topic and payload (Payload Format Indicator?). No MQTT v5 specific features such as user properties.
```

> * TCP multi-connection support (using Same Client ID)

Reviewer: This is impossible. One MQTT client can only have one TCP connection.
At a higher level, (e.g. a client SDK), maybe they can generate client IDs

Author: Impossible because we can't/won't implement it, or impossible because it's a fundamental limitations of the protocol?

> * User could manipulate file names, and paths. (blacklist/check file names/path, or relative path only)

Reviewer: Suggestion: 
* First publish: `$file/sha256sum/init` to initiate the transfer, and the payload can be a json body of file metadata (whatever the user wants to put in)
* Following messages: `$file/sha256sum/0` as topic name, payload for file content.

Reviewer: > Suggestion:
> 
> * First publish: `$file/sha256sum/init` to initiate the transfer, and the payload can be a json body of file metadata (whatever the user wants to put in)
> * Following messages: `$file/sha256sum/<byte-offset>` as topic name, payload for file content.

We will actually need a protocol for this init message in order to communicate the total size of the file etc.

Author: As far as I understand there are use cases when client would not know file size in advance, like rolling log file.

Author: How would broker deal with file metadata? Is it going to be a separate .json with the same name?

* Symmetric transfer (client -> broker)

> * PUBACK (QoS) guarantees transmission reliability (considering that the client does not receive PUBACK and resends it after disconnection)?

Reviewer: * Only allow `QoS=1`, PUBACK with error code can be used as failure indication.

Retransmission is already covered above.

| 0x83        | Implementation specific error | PUBACK | Receiver asks to cancel the transfer                |
| 0x97        | Quota exceeded                | PUBACK | Receiver asks to pause the transfer. Retry logic shall be agreed/configured before hand |

## Configuration Changes

This section should list all the changes to the configuration files (if any).

## Backwards Compatibility

This sections should shows how to make the feature is backwards compatible.
If it can not be compatible with the previous emqx versions, explain how do you
propose to deal with the incompatibilities.

## Document Changes

If there is any document change, give a brief description of it here.

## Testing Suggestions

The final implementation must include unit test or common test code. If some
more tests such as integration test or benchmarking test that need to be done
manually, list them here.

## Innovation Opportunities

* Integration with rule engine
* Support for more back-end, database writes
* Multiple backend writes to a single file
* Support for rules processing of metadata
* Generate modified properties of target files based on file metadata and rules of rules engine

> * Trigger event data when file upload starts/completes Enter rule chain for rule engine

Reviewer 3: ```suggestion
* Trigger event data when file upload starts/completes/enter rule chain for rule engine
```
?

* Server MUST be able to verify integrity of each file segment and of the whole file
* Client MAY know total file size when initiating the transfer
* Client MAY abort file transfer
* Server MAY ask the client to pause file transfer
* Server MAY ask the client to abort file transfer
* The protocol MUST guarantee "At least once" delivery
* Server MUST NOT support subscription on topics dedicated for file transfer

## AWS IoT MQTT-based file delivery (reference design)

As an example of existing implementation we can look at AWS IoT Core [which provides functionality](https://docs.aws.amazon.com/iot/latest/developerguide/mqtt-based-file-delivery.html) to [deliver files to IoT devices](https://docs.aws.amazon.com/iot/latest/developerguide/mqtt-based-file-delivery-in-devices.html):

![](0021-assets/aws-mqtt-file-delivery.png)

## Design

### Overview

* Files are split in segments, segments can be of arbitrary length
* Client MUST generate UUID according to [RFC 4122](https://www.rfc-editor.org/rfc/rfc4122) for each file being transferred and use it as file Id in Topic Name
* Broker MUST validate that provided UUID conforms to [RFC 4122](https://www.rfc-editor.org/rfc/rfc4122)

> * Client calculates sha256 checksum of the segment it's about to send and sends it as part of Topic Name

Reviewer: We should make this optional.
yesterday we discussed that doing sha256 checksum is quite an expensive thing.

* Broker MUST calculate checksum of the segment data it received and verify that it matches the one in Topic Name
* Client uses $file Topic to transfer files
* Broker MUST NOT let clients subscribe to $file topics
* Segment length can be calculated on the server side by subtracting the length of the [Variable Header](https://docs.oasis-open.org/mqtt/mqtt/v5.0/os/mqtt-v5.0-os.html#_Toc3901025) from the [Remaining Length](https://docs.oasis-open.org/mqtt/mqtt/v5.0/os/mqtt-v5.0-os.html#_Toc3901105) field that is in the [Fixed Header](https://docs.oasis-open.org/mqtt/mqtt/v5.0/os/mqtt-v5.0-os.html#_Toc3901021)
* Data is transferred in PUBLISH packets in the following order:
  1. $file/{fileId}/init
  2. $file/{fileId}/{offset}/{sha256sum}
  3. $file/{fileId}/{offset}/{sha256sum}
  4. ...
  3. $file/{fileId}/[fin/{sha256sum} | abort]

#### `$file/{fileId}/init` message

Initialize the file transfer. Server is expected to store metadata from the payload in the session along with `{fileId}` as a reference for the rest of file metadata.

  * Qos=1

>   * DUP=0 for the initial segment transfer, 1 when retransmitting

Reviewer: Not sure how useful is the DUP flag.
Re-sedning a init message for an existing ID should result in a purge of old file (or segments if it was not complete).

  * Payload Format Indicator=0x01
  * `{fileId}` is corresponding file UUID
  * Payload is a JSON document

##### `init` payload JSON Schema

Available [here](https://github.com/emqx/mqtt-file-transfer-schema/blob/v1.0.0/init.json).

* Broker SHOULD have default setting for `segments_ttl`

> * Broker MAY delete segments of unfinished file transfers when their TTL has expired

Reviewer: We should probably add an optional field to hold the assembled sha256 checksum.
And make the per-segment checksum optional too.

## Changelog

* 2022-09-20: @qzhuyan, @id Initial draft

## Abstract

TODO:

## Motivation

EMQX customers are asking for file transfer functionality from IoT devices to their cloud (primary use case), and from cloud to IoT devices over MQTT. Right now they are uploading files from devices via FTP or HTTPS (e.g. to S3), but this approach has downsides:

* FTP and HTTP servers usually struggle to keep up with large number of simultaneous bandwidth-intensive connections
* packet loss or reconnect forces clients to restart the transfer
* devices which already talk MQTT need to integrate with one more SDK, address authenticaion and authorization, and potentially go through an additional round of security audit

Known cases of device-to-cloud file transfer:

* [CAN bus](https://en.wikipedia.org/wiki/CAN_bus) data
* Image taken by industry camera for Quality Assurance
* Large data file collected from forklift
* Video and audio data from truck cars, and video data captured by inbound unloading cameras
* Vehicle real-time logging, telemetry, messaging
* Upload collected ML logs

Known cases of cloud-to-device file transfer:

* Upload AI/ML models
* Firmware upgrades

## Terminology

> The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "NOT RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be interpreted as described in [BCP 14](https://www.rfc-editor.org/bcp/bcp14) [[RFC2119](https://www.rfc-editor.org/rfc/rfc2119)] [[RFC8174](https://www.rfc-editor.org/rfc/rfc8174)] when, and only when, they appear in all capitals, as shown here.

Author: Not using these everywhere yet, and when I do, not always correctly. Will fix.

The following terms are used as described in [MQTT Version 5.0 Specification](https://docs.oasis-open.org/mqtt/mqtt/v5.0/os/mqtt-v5.0-os.html#_Toc39010030):
* Application Message
* Server
* Client
* Topic Name
* Topic Filter
* MQTT Control Packet

*At least once*: a message can be delivered many times, but cannot be lost

## Requirements

* The protocol MUST use only PUBLISH type of MQTT Control Packet
* The protocol MUST support transfer of file segments

> * Server MUST be able to verify integrity of each file segment

Reviewer: Is there a way to verify the integrity after putting all segments together?
Everything ends, right? even if it's a stream.

Author: final sha256 as part of the `fin` message? wdyt?

* Client MAY know total file size when initiating the transfer
* Client MAY abort file transfer
* Server MAY ask the client to pause file transfer
* Server MAY ask the client to abort file transfer

> * The protocol MUST NOT require changes in client code

Reviewer: I find it odd to put this here.  Maybe one requirement is that this portocol is on top of MQTT (which implies this is not imposing changes any existing MQTT client SDK). 
However I still find it odd to put implementation options in a protocol specification.
```suggestion

```

* The protocol MUST guarantee "At least once" delivery
* Server MUST NOT support subscription on topics dedicated for file transfer

## AWS IoT MQTT-based file delivery (reference design)

As an example of existing implementation we can look at AWS IoT Core [which provides functionality](https://docs.aws.amazon.com/iot/latest/developerguide/mqtt-based-file-delivery.html) to [deliver files to IoT devices](https://docs.aws.amazon.com/iot/latest/developerguide/mqtt-based-file-delivery-in-devices.html):

![](0021-assets/aws-mqtt-file-delivery.png)

## Design

### Overview

> * Files are split in segments of equal length with the exception of the last segment, it's length is > 0 and  <= segment length

Author: equal length is actually not critical, but if we don't do that, then we need to send segment offset together with payload, and the built-in MQTT mechanism of connecting PUBACK to corresponding PUBLISH messages via Packet Identifier might not always work, since Packet Identifier field is only 2 bytes.

Reviewer: 1. Let's not paint ourselve in this the corner or fixed size packet.
2. Let's avoid using Packet ID for other purposes than coorelating sent/ack (which is at MQTT layer, and we are on hihger layer). 
  * When using an SDK, packet ID is very much likely abstracted away.
  * For app layer controlled re-transmision, the same segment can be sent with two different packet IDs
  * The packet ID goes back to 0 after overflow
  * The packet ID can be set back to zero after idling in some implementations

> * Client generates UUID for each file being transferred and use it as file Id in Topic Name

Author: Not sure it's the right approach.
My thinking was as follows:
* we need some kind of unique file identifier anyway
* file path + file name could be fairly large, sending them as part of PUBLISH message with each segment could be greedy
* UUID can be stored in the session and broker can use it to get access to the rest of file metadata

* Client calculates sha256 checksum of the segment it's about to send and sends it as part of Topic Name
* Client uses $file Topic Filter to transfer files

> * Clients cannot subscribe to $file topics

Author: this implicitly tells that we do not support 'proxy' mode for file transfers

* Segment length can be calculated on the server side by subtracting the length of the [Variable Header](https://docs.oasis-open.org/mqtt/mqtt/v5.0/os/mqtt-v5.0-os.html#_Toc3901025) from the [Remaining Length](https://docs.oasis-open.org/mqtt/mqtt/v5.0/os/mqtt-v5.0-os.html#_Toc3901105) field that is in the [Fixed Header](https://docs.oasis-open.org/mqtt/mqtt/v5.0/os/mqtt-v5.0-os.html#_Toc3901021)
* Data is transferred in PUBLISH packets in the following order:
  1. $file/{fileId}/init
  2. $file/{fileId}/{sha256sum}
  3. $file/{fileId}/{sha256sum}
  4. ...
  3. $file/{fileId}/[fin|abort]

> * Client can send up to N (configured) PUBLISH packets before blocking for PUBACK

Author: TODO: remove this, MQTT clients have built-in in-flight window config

#### `$file/{fileId}/init` message

Initialize the file transfer. Server is expected to store metadata from the payload in the session along with `{fileId}` as a reference for the rest of file metadata.

  * Qos=1
  * DUP=0 for the initial segment transfer, 1 when retransmitting
  * Payload Format Indicator=0x01
  * `{fileId}` is corresponding file UUID
  * Payload is a JSON document with the following properties:
    * filepath
      * a character string composed with characters from this set [a-zA-Z0-9_-./]
      * length in bytes MUST be <= 4096 (PATH_MAX=4096 on most common Linux filesystems)
    * filename
      * a character string composed with characters from this set [a-zA-Z0-9_-./]
      * length in bytes MUST be < 256 (NAME_MAX=255 in Linux)

>     * size

Author: this can be optional

All file segments have been successfully transferred.

  * Qos=1
  * no payload

#### `$file/{fileId}/abort` message

Client wants to abort the transfer.

  * Qos=1
  * no payload

### PUBACK Reason codes

| Reason code | MQTT Name                     | Meaning in file transfer context                    |
|-------------|-------------------------------|-----------------------------------------------------|
| OMIT        |                               | Same as 0x00                                        |
| 0x00        | Success                       | File segment has been successfully persisted        |
| 0x10        | No matching subscribers       | Server asks Client to retransmit all segments       |
| 0x80        | Unspecified error             | Server asks Client to retransmit a specific segment. Segment sequential number is indicated by Packet Identifier field in the PUBACK Variable Header |
| 0x83        | Implementation specific error | Server asks Client to cancel the transfer           |
| 0x97        | Quota exceeded                | Server asks Client to pause the transfer            |

#### 0x83, "Implementation specific error", "Cancel Transfer"

> Client can retry the transfer after 1 hour.

Author: arbitrary number, open for suggestions

#### 0x97, "Quota exceeded", "Pause Transfer"

> Upon receiving PUBACK message with this reason code, Client is expected to delay sending next packet for 10 seconds. If next PUBACK is also 0x97, Client delays for 20 seconds. Client continues to double the delay length until it reaches 80 seconds.

Author: arbitrary numbers again, suggestions are welcome

* Clients cannot subscribe to $file topics
* Segment length can be calculated on the server side by subtracting the length of the [Variable Header](https://docs.oasis-open.org/mqtt/mqtt/v5.0/os/mqtt-v5.0-os.html#_Toc3901025) from the [Remaining Length](https://docs.oasis-open.org/mqtt/mqtt/v5.0/os/mqtt-v5.0-os.html#_Toc3901105) field that is in the [Fixed Header](https://docs.oasis-open.org/mqtt/mqtt/v5.0/os/mqtt-v5.0-os.html#_Toc3901021)
* Data is transferred in PUBLISH packets in the following order:
  1. $file/{fileId}/init
  2. $file/{fileId}/{offset}/{sha256sum}
  3. $file/{fileId}/{offset}/{sha256sum}
  4. ...
  3. $file/{fileId}/[fin/{sha256sum} | abort]
* Client can send up to N (configured) PUBLISH packets before blocking for PUBACK

#### `$file/{fileId}/init` message

Initialize the file transfer. Server is expected to store metadata from the payload in the session along with `{fileId}` as a reference for the rest of file metadata.

  * Qos=1
  * DUP=0 for the initial segment transfer, 1 when retransmitting
  * Payload Format Indicator=0x01
  * `{fileId}` is corresponding file UUID
  * Payload is a JSON document with the following properties:
    * filepath

>       * a character string composed with characters from this set [a-zA-Z0-9_-./]

Reviewer 4: What is the reason for the character set to be limited?

Author: TODO: remove this requirement

      * length in bytes MUST be <= 4096 (PATH_MAX=4096 on most common Linux filesystems)
    * filename

>       * a character string composed with characters from this set [a-zA-Z0-9_-./]

Author: TODO: remove this requirement

>       * length in bytes MUST be < 256 (NAME_MAX=255 in Linux)

Author: TODO: remove this requirement

    * size [optional]
      * total file size in bytes

#### `$file/{fileId}/{offset}/{sha256sum}` message

One such message for each file segment.

  * Qos=1
  * DUP=0 for the initial segment transfer, 1 when retransmitting
  * Payload Format Indicator=0x00

>   * Packet Identifier=sequential file segment number

Reviewer 4: Does this comply with the protocol? If we send several files, packet identifier will duplicate?

Author: TODO: remove this requirement

  * Payload is file segment bytes
  * `{offset}` is byte offset of the given segment

>   * `{sha256sum}` is sha256 checksum of file segment bytes

Author: * can we have offset and sha256 as part of the payload?
** upside: static and short topic names
** upside: clients can use topic alias if they support it
** downside: clients have to assemble payload, we need a serialization format

#### `$file/{fileId}/fin/{sha256sum}` message

All file segments have been successfully transferred.

  * Qos=1
  * no payload
  * `{sha256sum}` is sha256 checksum of the file

#### `$file/{fileId}/abort` message

Client wants to abort the transfer.

  * Qos=1
  * no payload

### PUBACK Reason codes

| Reason code | MQTT Name                     | Meaning in file transfer context                    |
|-------------|-------------------------------|-----------------------------------------------------|
| OMIT        |                               | Same as 0x00                                        |
| 0x00        | Success                       | File segment has been successfully persisted        |

> | 0x10        | No matching subscribers       | Server asks Client to retransmit all segments       |

Author: * could be used when e.g. file retention kicks in while client is still trying to send some later segments

* Broker MUST NOT let clients subscribe to $file topics
* Segment length can be calculated on the server side by subtracting the length of the [Variable Header](https://docs.oasis-open.org/mqtt/mqtt/v5.0/os/mqtt-v5.0-os.html#_Toc3901025) from the [Remaining Length](https://docs.oasis-open.org/mqtt/mqtt/v5.0/os/mqtt-v5.0-os.html#_Toc3901105) field that is in the [Fixed Header](https://docs.oasis-open.org/mqtt/mqtt/v5.0/os/mqtt-v5.0-os.html#_Toc3901021)
* Data is transferred in PUBLISH packets in the following order:
  1. $file/{fileId}/init
  2. $file/{fileId}/{offset}/{sha256sum}
  3. $file/{fileId}/{offset}/{sha256sum}
  4. ...
  3. $file/{fileId}/[fin/{sha256sum} | abort]

#### `$file/{fileId}/init` message

Initialize the file transfer. Server is expected to store metadata from the payload in the session along with `{fileId}` as a reference for the rest of file metadata.

  * Qos=1
  * DUP=0 for the initial segment transfer, 1 when retransmitting
  * Payload Format Indicator=0x01
  * `{fileId}` is corresponding file UUID
  * Payload is a JSON document

##### `init` payload JSON Schema

```json
{
  "$schema": "https://json-schema.org/draft/2020-12/schema",

>   "$id": "https://emqx.io/file-over-mqtt.init.schema.json",

Reviewer: we should put this to a git repo, and put a git tag on it, then reference it from here?

Author: I suck with naming. If you have a better idea what to call https://github.com/emqx/mqtt-file-transfer-schema, let me know.

  "title": "init",
  "description": "Schema for payload of the init message",
  "type": "object",
  "properties": {
    "name": {

>       "description": "File name",

Reviewer: ```suggestion
      "description": "A descriptive name for display",
```

      "type": "string"
    },
    "size": {
      "description": "Total file size in bytes, optional.",
      "type": "integer"
    },
    "expire_at": {
      "description": "File expiration time as UNIX Epoch.
                      After this point in time broker can delete the file.",
      "type": "integer"
    },

>     "segment_ttl": {

Reviewer: ```suggestion
    "segments_ttl": {
```

>       "description": "Time-to-leave (in seconds) for segments of incomplete file transfer.
>                       Broker will garbage collect segments of unfinished transfers when
>                       this time expires. If not set, broker will use its own default value.",

Reviewer: ```suggestion
      "description": "Time-to-live (in seconds) for segments of incomplete file transfer.
                      The deadline of for broker to collect all segments is calculated from InitTime + TTL,
                      where InitTime is the time when broker receives the init message. 
                      Broker will garbage collect segments of unfinished transfers when
                      this time expires. If not set, broker will use its own default value.",
```

      "type": "integer"
    },
    "user_data": {
      "description": "User-defined metadata, optional.",
      "type": "object"
    }
  },
  "required": [ "name", "expire_at" ]
}
```

> * Broker SHOULD have default setting for `segment_ttl`

Reviewer: ```suggestion
* Broker SHOULD have default setting for `segments_ttl`
```
